data-products/src/freestar_revenue_reporting/freestar_reporting_flow.py

```diff
@@ -0,0 +1,286 @@
+            "total": True,
+            "measures": [
+                "NdrPrebid.impressions",
+                "NdrPrebid.net_revenue",
+                "NdrPrebid.net_cpm",
+            ],
+            "dimensions": [
+                "NdrPrebid.record_date",
+                "NdrPrebid.network",
+                "NdrPrebid.url",
+                "NdrPrebid.utm_campaign",
+                "NdrPrebid.utm_content",
+                "NdrPrebid.utm_medium",
+                "NdrPrebid.utm_source",
+                "NdrPrebid.utm_term",
+                "NdrPrebid.site_domain",
+                "NdrPrebid.ad_unit",
+                "NdrPrebid.size",
+                "NdrPrebid.country_code",
+                "NdrPrebid.device_type",
+                "NdrPrebid.device_os",
+                "NdrPrebid.browser",
+                "NdrPrebid.integration_partner",
+            ],
+            "timeDimensions": [
+                {
+                    "dimension": "NdrPrebid.record_date",
+                    "dateRange": "Yesterday",  # for backfill use "dateRange": ["2023-09-01", "2023-09-08"]
+                }
+            ],
+            "limit": 2000,  # Retrieve 2,000 records at a time to accomodate 5 second timeout
+            "offset": 0,  # Start with an offset of 0
+        }
+    }
+
+    # Set up the headers with the API token
+    headers = {
+        "Authorization": f"Bearer {FREESTAR_API_KEY}",
+        "Content-Type": "application/json",
+    }
+
+    page_number = 1  # Initialize the page number tracker
+    total_count = 0  # Initialize the total count tracker
+
+    # create directory
+    Path(output_parquet_path).mkdir(parents=True, exist_ok=True)
+
+    while True:
+        # Send the POST request to the API with the current pagination parameters
+        response = requests.post(
+            API_BASE_URL + API_ENDPOINT, json=request_payload, headers=headers
+        )
+        response.raise_for_status()  # Raise an exception for HTTP errors
+
+        # Parse the JSON response
+        response_json = response.json()
+        logger.info(
+            f"Retrieved {len(response_json['data'])} records for page {page_number}."
+        )
+
+        # Remove the 'NdrPrebid' prefix from keys in the JSON response
+        for item in response_json["data"]:
+            for key in list(item.keys()):
+                if key.startswith("NdrPrebid."):
+                    new_key = key[len("NdrPrebid.") :]
+                    item[new_key] = item.pop(key)
+
+        # Write to file
+        data = response_json["data"]
+        record_count = len(data)
+        df = pd.DataFrame(data)
+        df.to_parquet(output_parquet_filename.format(page_number))
+        total_count += record_count
+
+        # Check if there are more records to retrieve
+        if record_count < 2000:
+            break  # Stop if there are no more records
+
+        # Increment the offset for the next page
+        request_payload["query"]["offset"] += 2000
```

Comment:
Qs:
- will there ever be a situation where record_count > 2000?
- would it make a difference to put `+= record_count` instead of hard-coding 2000?
- would `record_count < 2000` always mean the last fetch?
- would there be an advantage to set the page size as a variable if there is a reason to make adjustments, instead of the hard-coded value of 2000?

Reply:
@gkatre the 2,000 count shouldn't change. Freestar has a 5-second timeout on their API, so we can't pull many records at a time. 2,000 was the sweet spot. I tested 2,500, 3,000, & 5,000. The higher we go the more frequently it fails. 2,000 records at a time results in the least amount of retries (I haven't seen it retry more than once, but set retries to 3 to be safe). I suppose Freestar could increase their API timeout in the future and if we were made aware we could test it all again, find the sweet spot for the number of records, and change this value.

But, you're right that it should be stored in a variable to keep things DRY. Thanks, that is a good point and I have pushed that change. :) 

Yes, record_count < 2000 always means the last fetch...I've tested this extensively - compared to manually downloading csv file for record count